packages/tpetra/core/src/Tpetra_Filter.hpp

```diff
@@ -0,0 +1,216 @@
+#ifndef TPETRA_FILTER_HPP
+#define TPETRA_FILTER_HPP
+
+#include "Kokkos_Core.hpp"
+#include "Teuchos_RCP.hpp"
+
+namespace Tpetra {
+
+/// \brief Constructs a matrix consisting only of the entries for which the filter function returns true.
+///
+/// This applies a filter function that takes as single argument the value of the matrix entry.
+template <class matrix_type, class filter_type>
```

Comment:
Give this a doxygen long form that describes arguments to filter (in order) for each case, with associated data types.